components/places/tests/fennec_bookmarks.rs

```diff
@@ -466,3 +459,139 @@ fn test_empty_db() -> Result<()> {
+    } else {
+        (now - Duration::new(20, 0)).into()
+    };
+
+    let bookmarks = [FennecBookmark {
+        _id: 6,
+        guid: "bookmarkAAAA".into(),
+        created: Some(timestamp_local),
+        modified: Some(timestamp_local),
+        position: 0,
+        parent: 5,
+        // It doesn't matter what the tag is, just that *something* exists,
+        // because our SQL will force an early modified date if it does.
+        tags: Some("whatever".to_owned()),
+        title: Some("A".to_owned()),
+        url: Some("http://example.com/a/%s".to_owned()),
+        ..Default::default()
+    }];
+    insert_bookmarks(&fennec_db, &get_fennec_roots())?;
+    insert_bookmarks(&fennec_db, &bookmarks)?;
+
+    let places_api = PlacesApi::new(tmpdir.path().join("places.sqlite"))?;
+    let metrics = places::import::import_fennec_bookmarks(&places_api, fennec_path)?;
+    assert_eq!(metrics.num_failed, 0);
+
+    let writer = places_api.open_connection(ConnectionType::ReadWrite)?;
+    let syncer = places_api.open_sync_connection()?;
+
+    // Should be no bookmark with keyword 'a' yet.
+    assert_eq!(bookmarks_get_url_for_keyword(&writer, "a")?, None);
+    // And no URL with our test tag yet.
+    assert_eq!(tags::get_urls_with_tag(&writer, "test-tag")?, []);
+
+    // Now setup incoming records from the server with some of the data we
+    // missed.
+    let records = vec![
+        json!({
+            "id": "unfiled",
+            "type": "folder",
+            "parentid": "places",
+            "parentName": "root",
+            "dateAdded": DEFAULT_BOOKMARK_CREATED_TIMESTAMP,
+            "title": "unfiled",
+            "children": ["bookmarkAAAA"],
+        }),
+        json!({
+            "id": "bookmarkAAAA",
+            "type": "bookmark",
+            "parentid": "unfiled",
+            "parentName": "unfiled",
+            "dateAdded": DEFAULT_BOOKMARK_CREATED_TIMESTAMP,
+            "title": "A",
+            "bmkUri": "http://example.com/a/%s",
+            "keyword": "a",
+            "tags": ["test-tag"],
+        }),
+    ];
+
+    let interrupt_scope = syncer.begin_interrupt_scope();
+    let store = BookmarksStore::new(&syncer, &interrupt_scope);
+
+    let mut incoming =
+        IncomingChangeset::new(store.collection_name().to_string(), server_timestamp);
+    for record in records {
+        let payload = Payload::from_json(record).unwrap();
+        incoming.changes.push((payload, item_server_timestamp));
+    }
+
+    let outgoing = store
+        .apply_incoming(incoming, &mut telemetry::Engine::new("bookmarks"))
+        .expect("Should apply incoming records");
+    let outgoing_records: HashMap<_, _> =
+        outgoing.changes.iter().map(|p| (p.id.clone(), p)).collect();
+
+    println!("outgoing: {:#?}", outgoing_records);
+    // XXX - All roots are in outgoing - including 'unfiled' which was in
+    // incoming - not quite clear why, although I note dateAdded is wrong in
+    // every case.
```

Comment:
This is the case we talked about on Slack. Dogear makes a [special exception for roots](https://github.com/mozilla/dogear/blob/0b782ddf376bb4c42b2a039b66ec1699b88feb8f/src/merge.rs#L1002-L1006). If the local root has `sync_change_counter > 0`, we'll always reupload it, even if we used the order of `children` from the server. This is because we always want to use our local titles, which are now hard-coded constants instead of strings. The Desktop bug for this landed [almost 2 years ago](https://bugzilla.mozilla.org/show_bug.cgi?id=1432614), so we can probably remove this special-casing from Dogear. But, as you said, reuploading the roots is otherwise harmless, and we don't want to introduce new changes this late.